### PR TITLE
ddl, expression: fix some string functions' flen and refine error message for expression index

### LINKED ISSUE #27296: `EXPORT_SET` result type incompatible with MySQL  lead to create index failed
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
create table t1(a char, b varchar(20), c char, d varchar(20));
alter table t1 add index eidx ((export_set(3, a, c, ',', 5)));

```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
```sql
mysql> alter table t1 add index eidx ((export_set(3, a, c, ',', 5)));
Query OK, 0 rows affected (0.05 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> SELECT EXPORT_SET(5,'Y','N',',',4);
Field   1:  `EXPORT_SET(5,'Y','N',',',4)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8mb4_general_ci (45)
Length:     508
Max_length: 7
Decimals:   31
Flags:


```
### 3. What did you see instead (Required)
```sql
mysql> alter table t1 add index eidx ((export_set(3, a, c, ',', 5)));
ERROR 1170 (42000): BLOB/TEXT column '_V$_eidx_0' used in key specification without a key length

mysql> SELECT EXPORT_SET(5,'Y','N',',',4);
Field   1:  `EXPORT_SET(5,'Y','N',',',4)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       LONG_BLOB
Collation:  utf8mb4_bin (46)
Length:     67108864
Max_length: 7
Decimals:   31
Flags:      NOT_NULL



```
### 4. What is your TiDB version? (Required)
```sql
| tidb_version()                                                                                                                                                                                                                                                                                                       |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v5.2.0
Edition: Community
Git Commit Hash: 88f60fe1140de999c2e64ca56edffb7db7393f1e
Git Branch: heads/refs/tags/v5.2.0
UTC Build Time: 2021-08-13 15:08:56
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```
<!-- Paste the output of SELECT tidb_version() -->

### LINKED ISSUE #27361: query and insert data error whlie `concat_ws` as expression index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
create table t01(a varchar(20));
insert into t01 values ("齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙");
alter table t01 add index eidx ((concat_ws('expression_index', a, 'test')));
 select * from t01 use index (eidx)  where (concat_ws('expression_index', a, 'test')) not like (concat_ws('expression_index', "齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙", 'test'));
insert into t01 values ("齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙");
```

### 2. What did you expect to see? (Required)
```sql
MySQL root@172.16.5.103:hchwang> select * from t01 use index (eidx)  where (concat_ws('expression_index', a, 'test')) not like (concat_ws('expression_index', "齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙", 'test'));
+---+
| a |
+---+
0 rows in set
Time: 0.054s
MySQL root@172.16.5.103:hchwang> insert into t01 values ("齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙");
Query OK, 1 row affected
Time: 0.039s
MySQL root@172.16.5.103:hchwang> select * from t01;
+------------------------------------------+
| a                                        |
+------------------------------------------+
| 齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙 |
| 齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙 |
+------------------------------------------+
```
### 3. What did you see instead (Required)
```sql
MySQL root@127.0.0.1:test> drop table t01;
Query OK, 0 rows affected
Time: 0.019s
MySQL root@127.0.0.1:test> create table t01(a varchar(20));
Query OK, 0 rows affected
Time: 0.011s
MySQL root@127.0.0.1:test> insert into t01 values ("齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙");
Query OK, 1 row affected
Time: 0.001s
MySQL root@127.0.0.1:test> alter table t01 add index eidx ((concat_ws('expression_index', a, 'test')));
Query OK, 0 rows affected
Time: 2.547s

MySQL root@127.0.0.1:test> select * from t01 use index (eidx)  where (concat_ws('expression_index', a, 'test')) not like (concat_ws('expression_index', "齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙", 'test'));
+------------------------------------------+
| a                                        |
+------------------------------------------+
| 齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙 |
+------------------------------------------+
1 row in set
Time: 0.008s
MySQL root@127.0.0.1:test> insert into t01 values ("齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙");
(1406, 'Data Too Long, field len 25, data len 40')

MySQL root@127.0.0.1:test> select * from t01;
+------------------------------------------+
| a                                        |
+------------------------------------------+
| 齆斮聒蚆髙锐潊贩哨啅捸爖斥圱犳飁綴纜牖蚙 |
+------------------------------------------+

```
### 4. What is your TiDB version? (Required)

```sql 
tidb_version() | Release Version: v5.2.0-alpha-567-gdba8e111f-dirty
Edition: Community
Git Commit Hash: dba8e111fe26b1db0742dd5a024309137d49f5d2
Git Branch: release-5.2
UTC Build Time: 2021-08-19 01:23:29
GoVersion: go1.16.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #27296, close #27361 

Problem Summary:
These functions' flen are wrong.

### What is changed and how it works?

For exportSet(), set it flen to 508(the same as MySQL)
For concatWS(), consider the flen of the first argument, make it the same as concat(). 
For fromBase64(), calculate it flen.

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
'None'
```

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -3073,7 +3073,10 @@ func (c *exportSetFunctionClass) getFunction(ctx sessionctx.Context, args []Expr
 	if err != nil {
 		return nil, err
 	}
-	bf.tp.Flen = mysql.MaxBlobWidth
+	// Max length is (64 * 2 -1) * 4
+	// For example:
+	// "Y,Y,Y,Y,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N,N"
+	bf.tp.Flen = 508
```
Comment (wshwsh12) [inline expression/builtin_string.go:3079]: What will happen if arg[1]/arg[2] is too long? 
I think using the formula in mysql is better.. https://github.com/mysql/mysql-server/blob/8.0/sql/item_strfunc.cc#L3646

Comment (wjhuang2016) [reply expression/builtin_string.go:3079]: Any example？

Comment (wjhuang2016) [reply expression/builtin_string.go:3079]: It seems that MySQL always return 508.
``` sql
mysql> SELECT EXPORT_SET(5,'Y','N',',',4);
Field   1:  `EXPORT_SET(5,'Y','N',',',4)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8mb4_0900_ai_ci (255)
Length:     508
Max_length: 7
Decimals:   31
Flags:


+-----------------------------+
| EXPORT_SET(5,'Y','N',',',4) |
+-----------------------------+
| Y,N,Y,N                     |
+-----------------------------+
1 row in set (0.00 sec)
```

Comment (wjhuang2016) [reply expression/builtin_string.go:3079]: OK, I will fix it.

File: errno/errcode.go [hunk 2]
```diff
@@ -891,7 +891,7 @@ const (
 	ErrFunctionalIndexRefAutoIncrement                       = 3754
 	ErrCannotDropColumnFunctionalIndex                       = 3755
 	ErrFunctionalIndexPrimaryKey                             = 3756
-	ErrFunctionalIndexOnLob                                  = 3757
+	ErrFunctionalIndexOnBlob                                 = 3757
```
Comment (wshwsh12) [inline errno/errcode.go:894]: Should we add the error into errors.toml?

**Review by wshwsh12 — COMMENTED** [review]
Rest LGTM

**Review by wshwsh12 — APPROVED** [review]

File: errno/errcode.go (hunk 2 above)
Comment (wjhuang2016) [reply errno/errcode.go:894]: No, I tried `make errdoc` already.

File: expression/builtin_string.go [hunk 3]
```diff
@@ -3073,7 +3073,16 @@ func (c *exportSetFunctionClass) getFunction(ctx sessionctx.Context, args []Expr
...
+	// Calculate the flen as MySQL does.
+	l := args[1].GetType().Flen
+	if args[2].GetType().Flen > l {
+		l = args[2].GetType().Flen
+	}
+	sepL := 1
+	if len(args) > 3 {
+		sepL = args[3].GetType().Flen
+	}
+	bf.tp.Flen = (l*64 + sepL*63) * 4
```
Comment (AilinKid) [inline expression/builtin_string.go:3085]: any calculation reference link for this?

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

**Review by tangenta — APPROVED** [review]

File: expression/builtin_string.go (hunk 3 above)
Comment (wjhuang2016) [reply expression/builtin_string.go:3085]: Refer to @wshwsh12 's comment.

**Review by AilinKid — APPROVED** [review]

**Pull request merged**